File: Hard/2025-12-16-3562-Maximum-Profit-from-Trading-Stocks-with-Discounts/solution.js
class Solution {
    maxProfit(n, present, future, hierarchy, budget) {
        const graph = Array.from({ length: n + 1 }, () => []);
        for (const [u, v] of hierarchy) {
            graph[u].push(v);
        }

        const dp = Array.from({ length: budget + 1 }, () => 0);

        const dfs = (employee) => {
            const costs = [];
            const profits = [];
            costs.push(present[employee - 1]);
            profits.push(future[employee - 1] - present[employee - 1]);

            for (const subordinate of graph[employee]) {
                const [subCost, subProfit] = dfs(subordinate);
                costs.push(Math.floor(present[subordinate - 1] / 2));
                profits.push(future[subordinate - 1] - Math.floor(present[subordinate - 1] / 2));
            }

            const newDp = [...dp];
            for (let i = budget; i >= 0; i--) {
                for (let j = 0; j < costs.length; j++) {
                    if (i >= costs[j]) {
                        newDp[i] = Math.max(newDp[i], dp[i - costs[j]] + profits[j]);
                    }
                }
            }
            return [costs, profits];
        };

        dfs(1);
        return Math.max(...dp);
    }
}

File: Hard/2025-12-16-3562-Maximum-Profit-from-Trading-Stocks-with-Discounts/solution.py
class Solution:
    def maximumProfit(self, n: int, present: List[int], future: List[int], hierarchy: List[List[int]], budget: int) -> int:
        from collections import defaultdict
        import heapq
        
        graph = defaultdict(list)
        for u, v in hierarchy:
            graph[u].append(v)
        
        profits = []
        costs = []
        
        def dfs(employee):
            nonlocal profits, costs
            buy_cost = present[employee - 1]
            sell_price = future[employee - 1]
            profit = sell_price - buy_cost
            
            profits.append(profit)
            costs.append(buy_cost)
            
            for subordinate in graph[employee]:
                discounted_cost = present[subordinate - 1] // 2
                discounted_profit = future[subordinate - 1] - discounted_cost
                
                profits.append(discounted_profit)
                costs.append(discounted_cost)
                
                dfs(subordinate)
        
        dfs(1)
        
        max_profit = 0
        dp = [0] * (budget + 1)
        
        for cost, profit in zip(costs, profits):
            for b in range(budget, cost - 1, -1):
                dp[b] = max(dp[b], dp[b - cost] + profit)
        
        return max(dp)